RecoLocalCalo/HcalRecProducers/src/KernelHelpers.h

```diff
@@ -0,0 +1,230 @@
+
+    // compute the charge using the adc, qie type and the appropriate qie shape array
+    __forceinline__ __device__ float compute_coder_charge(
+        int const qieType, uint8_t const adc, uint8_t const capid, float const* qieOffsets, float const* qieSlopes) {
+      auto const range = qieType == 0 ? (adc >> 5) & 0x3 : (adc >> 6) & 0x3;
+      auto const* qieShapeToUse = qieType == 0 ? qie8shape : qie11shape;
+      auto const nbins = qieType == 0 ? 32 : 64;
+      auto const center = adc % nbins == nbins - 1 ? 0.5 * (3 * qieShapeToUse[adc] - qieShapeToUse[adc - 1])
+                                                   : 0.5 * (qieShapeToUse[adc] + qieShapeToUse[adc + 1]);
+      auto const index = get_qiecoder_index(capid, range);
+      return (center - qieOffsets[index]) / qieSlopes[index];
+    }
+
+    __forceinline__ __device__ float compute_diff_charge_gain(int const qieType,
+                                                              uint8_t adc,
+                                                              uint8_t const capid,
+                                                              float const* qieOffsets,
+                                                              float const* qieSlopes,
+                                                              bool const isqie11) {
+      constexpr uint32_t mantissaMaskQIE8 = 0x1fu;
+      constexpr uint32_t mantissaMaskQIE11 = 0x3f;
+      auto const mantissaMask = isqie11 ? mantissaMaskQIE11 : mantissaMaskQIE8;
+      auto const q = compute_coder_charge(qieType, adc, capid, qieOffsets, qieSlopes);
+      auto const mantissa = adc & mantissaMask;
+
+      if (mantissa == 0u || mantissa == mantissaMask - 1u)
+        return compute_coder_charge(qieType, adc + 1u, capid, qieOffsets, qieSlopes) - q;
+      else if (mantissa == 1u || mantissa == mantissaMask)
+        return q - compute_coder_charge(qieType, adc - 1u, capid, qieOffsets, qieSlopes);
+      else {
+        auto const qup = compute_coder_charge(qieType, adc + 1u, capid, qieOffsets, qieSlopes);
+        auto const qdown = compute_coder_charge(qieType, adc - 1u, capid, qieOffsets, qieSlopes);
+        auto const upgain = qup - q;
+        auto const downgain = q - qdown;
+        auto const averagegain = (qup - qdown) / 2.f;
+        if (std::abs(upgain - downgain) < 0.01f * averagegain)
+          return averagegain;
+        else {
+          auto const q2up = compute_coder_charge(qieType, adc + 2u, capid, qieOffsets, qieSlopes);
+          auto const q2down = compute_coder_charge(qieType, adc - 2u, capid, qieOffsets, qieSlopes);
+          auto const upgain2 = q2up - qup;
+          auto const downgain2 = qdown - q2down;
+          if (std::abs(upgain2 - upgain) < std::abs(downgain2 - downgain))
+            return upgain;
+          else
+            return downgain;
+        }
+      }
+    }
+
+    // FIXME remove duplication...
+    // this is from PulesFunctor. nvcc was complaining... if included that header...
```

Comment:
Out of curiosity: what is that `PulesFunctor`?
If `PulseShapeFunctor`, then please fix the comment, to help the reader